pkg/bloomgateway/worker.go

```diff
@@ -104,109 +105,118 @@ func (w *worker) starting(_ context.Context) error {
 	return nil
 }
 
-func (w *worker) running(ctx context.Context) error {
+func (w *worker) running(_ context.Context) error {
 	idx := queue.StartIndexWithLocalQueue
 
-	for {
-		select {
+	for st := w.State(); st == services.Running || st == services.Stopping; {
+		taskCtx := context.Background()
+		dequeueStart := time.Now()
+		items, newIdx, err := w.queue.DequeueMany(taskCtx, idx, w.id, w.cfg.maxItems, w.cfg.maxWaitTime)
+		w.metrics.dequeueWaitTime.WithLabelValues(w.id).Observe(time.Since(dequeueStart).Seconds())
+		if err != nil {
+			// We only return an error if the queue is stopped and dequeuing did not yield any items
+			if err == queue.ErrStopped && len(items) == 0 {
+				return err
+			}
+			w.metrics.dequeueErrors.WithLabelValues(w.id).Inc()
+			level.Error(w.logger).Log("msg", "failed to dequeue tasks", "err", err, "items", len(items))
+		}
+		idx = newIdx
 
-		case <-ctx.Done():
-			return ctx.Err()
+		if len(items) == 0 {
+			w.queue.ReleaseRequests(items)
+			continue
+		}
+		w.metrics.dequeuedTasks.WithLabelValues(w.id).Add(float64(len(items)))
 
-		default:
-			taskCtx := context.Background()
-			dequeueStart := time.Now()
-			items, newIdx, err := w.queue.DequeueMany(taskCtx, idx, w.id, w.cfg.maxItems, w.cfg.maxWaitTime)
-			w.metrics.dequeueWaitTime.WithLabelValues(w.id).Observe(time.Since(dequeueStart).Seconds())
-			if err != nil {
-				// We only return an error if the queue is stopped and dequeuing did not yield any items
-				if err == queue.ErrStopped && len(items) == 0 {
-					return err
-				}
-				w.metrics.dequeueErrors.WithLabelValues(w.id).Inc()
-				level.Error(w.logger).Log("msg", "failed to dequeue tasks", "err", err, "items", len(items))
-			}
-			idx = newIdx
+		tasksPerDay := make(map[model.Time][]Task)
 
-			if len(items) == 0 {
+		for _, item := range items {
+			task, ok := item.(Task)
+			if !ok {
+				// This really should never happen, because only the bloom gateway itself can enqueue tasks.
 				w.queue.ReleaseRequests(items)
-				continue
+				return errors.Errorf("failed to cast dequeued item to Task: %v", item)
 			}
-			w.metrics.dequeuedTasks.WithLabelValues(w.id).Add(float64(len(items)))
+			level.Debug(w.logger).Log("msg", "dequeued task", "task", task.ID)
+			w.pending.Delete(task.ID)
 
-			tasksPerDay := make(map[model.Time][]Task)
-
-			for _, item := range items {
-				task, ok := item.(Task)
-				if !ok {
-					// This really should never happen, because only the bloom gateway itself can enqueue tasks.
-					w.queue.ReleaseRequests(items)
-					return errors.Errorf("failed to cast dequeued item to Task: %v", item)
-				}
-				level.Debug(w.logger).Log("msg", "dequeued task", "task", task.ID)
-				w.pending.Delete(task.ID)
+			tasksPerDay[task.day] = append(tasksPerDay[task.day], task)
+		}
 
-				tasksPerDay[task.day] = append(tasksPerDay[task.day], task)
+		for day, tasks := range tasksPerDay {
```

Comment:
I think we should check early if the context is canceled here. Maybe the context was canceled while processing the previous day.

I see it's done after calling `GetBlockRefs` (which will return an err if the task ctx is closed) but I don't see why not doing it before.

Reply:
Agree, we can check at multiple places for cancelled context